test/caching_iterable_test.dart

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test/test.dart';
+
+import 'package:collection/collection.dart';
+
+int yieldCount;
+
+Iterable<int> range(int start, int end) sync* {
+  assert(yieldCount == 0);
+  for (int index = start; index <= end; index += 1) {
+    yieldCount += 1;
+    yield index;
+  }
+}
+
+void main() {
+  setUp(() {
+    yieldCount = 0;
+  });
+
+  test('The Caching Iterable: length caches', () {
+    final Iterable<int> i = new CachingIterable<int>(range(1, 5).iterator);
```

Comment:
`var iterable` or maybe `var integers` to match tests below.

The generic type in the constructor invocation shouldn't be necessary either.